File: public/proxy-server/scripts/proxy.py
#!/usr/bin/env python3
"""
API Client CORS Proxy Server (Python)

A lightweight proxy server to bypass CORS restrictions during API development.
Uses only Python standard library - no external dependencies required.
"""

import asyncio
import base64
import json
import argparse
import sys
import urllib.parse
import urllib.request
import urllib.error
import ssl
from http.server import HTTPServer, BaseHTTPRequestHandler
from http import HTTPStatus
import time

class ProxyConfig:
    def __init__(self):
        self.port = 9090
        self.host = '0.0.0.0'
        self.cors_origin = '*'
        self.username = None
        self.password = None
        self.ssl_verification = 'default'
        self.verbose = False

config = ProxyConfig()

class CORSProxyHandler(BaseHTTPRequestHandler):
    def log_message(self, format, *args):
        if config.verbose:
            super().log_message(format, *args)
    
    def check_auth(self):
        """Check basic authentication if configured"""
        if not config.username or not config.password:
            return True
        
        auth_header = self.headers.get('Proxy-Authorization') or self.headers.get('Authorization')
        if not auth_header:
            return False
        
        try:
            auth_type, credentials = auth_header.split(' ', 1)
            if auth_type.lower() != 'basic':
                return False
            
            decoded = base64.b64decode(credentials).decode('utf-8')
            username, password = decoded.split(':', 1)
            
            return username == config.username and password == config.password
        except:
            return False
    
    def send_cors_headers(self):
        """Add CORS headers to response"""
        origin = self.headers.get('Origin', '*')
        
        if config.cors_origin == '*':
            self.send_header('Access-Control-Allow-Origin', origin)
        elif isinstance(config.cors_origin, list):
            if origin in config.cors_origin:
                self.send_header('Access-Control-Allow-Origin', origin)
        else:
            self.send_header('Access-Control-Allow-Origin', config.cors_origin)
        
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, PUT, DELETE, PATCH, OPTIONS, HEAD')
        self.send_header('Access-Control-Allow-Headers', '*')
        self.send_header('Access-Control-Allow-Credentials', 'true')
        self.send_header('Access-Control-Max-Age', '86400')
    
    def do_OPTIONS(self):
        """Handle preflight requests"""
        self.send_response(HTTPStatus.NO_CONTENT)
        self.send_cors_headers()
        self.end_headers()
    
    def do_GET(self):
        self.handle_request()
    
    def do_POST(self):
        self.handle_request()
    
    def do_PUT(self):
        self.handle_request()
    
    def do_DELETE(self):
        self.handle_request()
    
    def do_PATCH(self):
        self.handle_request()
    
    def do_HEAD(self):
        self.handle_request()
    
    def handle_request(self):
        """Handle all HTTP methods"""
        # Handle health check
        if self.path == '/health':
            self.send_response(HTTPStatus.OK)
            self.send_header('Content-Type', 'application/json')
            self.send_cors_headers()
            self.end_headers()
            
            health_data = {
                'status': 'ok',
                'version': '1.0.0',
                'uptime': time.time() - start_time
            }
            self.wfile.write(json.dumps(health_data).encode())
            return
        
        # Check authentication
        if not self.check_auth():
            self.send_response(HTTPStatus.UNAUTHORIZED)
            self.send_header('WWW-Authenticate', 'Basic realm="Proxy"')
            self.send_cors_headers()
            self.end_headers()
            self.wfile.write(b'Unauthorized')
            return
        
        # Extract target URL
        target_url = self.headers.get('X-Target-URL')
        if not target_url:
            # Check query parameter
            parsed_path = urllib.parse.urlparse(self.path)
            query_params = urllib.parse.parse_qs(parsed_path.query)
            target_url = query_params.get('url', [None])[0]
        
        if not target_url:
            self.send_response(HTTPStatus.BAD_REQUEST)
            self.send_cors_headers()
            self.end_headers()
            self.wfile.write(b'Missing target URL. Use X-Target-URL header or ?url= parameter')
            return
        
        try:
            # Prepare request
            req = urllib.request.Request(target_url, method=self.command)
            
            # Copy headers
            skip_headers = {
                'host', 'connection', 'proxy-authorization',
                'x-target-url', 'origin', 'referer'
            }
            
            for header, value in self.headers.items():
                if header.lower() not in skip_headers:
                    req.add_header(header, value)
            
            # Add body if present
            content_length = self.headers.get('Content-Length')
            if content_length:
                body = self.rfile.read(int(content_length))
                req.data = body
            
            # Handle SSL verification
            if config.ssl_verification == 'ignore':
                ssl_context = ssl.create_default_context()
                ssl_context.check_hostname = False
                ssl_context.verify_mode = ssl.CERT_NONE
            else:
                ssl_context = None
            
            # Make request
            if config.verbose:
                print(f"[{time.strftime('%Y-%m-%d %H:%M:%S')}] {self.command} {target_url}")
            
            response = urllib.request.urlopen(req, context=ssl_context)
            
            # Send response
            self.send_response(response.code)
            
            # Copy response headers
            for header, value in response.headers.items():
                if header.lower() not in ('connection', 'transfer-encoding'):
                    self.send_header(header, value)
            
            self.send_cors_headers()
            self.end_headers()
            
            # Copy response body
            while True:
                chunk = response.read(8192)
                if not chunk:
                    break
                self.wfile.write(chunk)
            
            if config.verbose:
                print(f"[{time.strftime('%Y-%m-%d %H:%M:%S')}] {self.command} {target_url} -> {response.code}")
                
        except urllib.error.HTTPError as e:
            self.send_response(e.code)
            
            # Copy error response headers
            for header, value in e.headers.items():
                if header.lower() not in ('connection', 'transfer-encoding'):
                    self.send_header(header, value)
            
            self.send_cors_headers()
            self.end_headers()
            
            # Copy error body
            self.wfile.write(e.read())
            
            if config.verbose:
                print(f"[{time.strftime('%Y-%m-%d %H:%M:%S')}] {self.command} {target_url} -> {e.code}")
                
        except Exception as e:
            self.send_response(HTTPStatus.INTERNAL_SERVER_ERROR)
            self.send_header('Content-Type', 'application/json')
            self.send_cors_headers()
            self.end_headers()
            
            error_data = {
                'error': 'Proxy error',
                'message': str(e)
            }
            self.wfile.write(json.dumps(error_data).encode())
            
            if config.verbose:
                print(f"[{time.strftime('%Y-%m-%d %H:%M:%S')}] {self.command} {target_url} -> 500 (Error: {e})")

def main():
    parser = argparse.ArgumentParser(description='API Client CORS Proxy Server')
    parser.add_argument('-p', '--port', type=int, default=9090, help='Port to listen on (default: 9090)')
    parser.add_argument('-H', '--host', default='0.0.0.0', help='Host to bind to (default: 0.0.0.0)')
    parser.add_argument('--origin', default='*', help='CORS origin(s) to allow (default: *)')
    parser.add_argument('-u', '--username', help='Basic auth username')
    parser.add_argument('--password', help='Basic auth password')
    parser.add_argument('-v', '--verbose', action='store_true', help='Enable verbose logging')
    parser.add_argument('--ssl', choices=['default', 'ignore', 'strict'], default='default',
                       help='SSL verification mode')
    
    args = parser.parse_args()
    
    # Update config
    config.port = args.port
    config.host = args.host
    config.cors_origin = args.origin
    config.username = args.username
    config.password = args.password
    config.ssl_verification = args.ssl
    config.verbose = args.verbose
    
    # Start server
    global start_time
    start_time = time.time()
    
    server = HTTPServer((config.host, config.port), CORSProxyHandler)
    
    print(f"""
🚀 API Client CORS Proxy Server (Python) started!
   
   Listening on: http://{config.host}:{config.port}
   Health check: http://{config.host}:{config.port}/health
   
   Usage:
   - Add X-Target-URL header with the target API URL
   - Or use ?url=<target-url> query parameter
   
   {'🔒 Authentication enabled' if config.username else '🔓 No authentication'}
   {'📝 Verbose logging enabled' if config.verbose else ''}
""")
    
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print('\n👋 Shutting down proxy server...')
        server.shutdown()

if __name__ == '__main__':
    main()

File: proxy-server/scripts/proxy.go
package main

/*
API Client CORS Proxy Server (Go)

A lightweight proxy server to bypass CORS restrictions during API development.
No external dependencies - uses only Go standard library.

Usage:
  go run proxy.go [flags]

Flags:
  -port int      Port to listen on (default 9090)
  -host string   Host to bind to (default "0.0.0.0")
  -origin string CORS origin to allow (default "*")
  -username string Basic auth username
  -password string Basic auth password
  -verbose       Enable verbose logging
*/

import (
	"encoding/base64"
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"log"
	"net/http"
	"strings"
	"time"
)

type Config struct {
	Port     int
	Host     string
	Origin   string
	Username string
	Password string
	Verbose  bool
}

type HealthResponse struct {
	Status  string  `json:"status"`
	Version string  `json:"version"`
	Uptime  float64 `json:"uptime"`
}

type ErrorResponse struct {
	Error   string `json:"error"`
	Message string `json:"message"`
}

var (
	config    Config
	startTime time.Time
)

func main() {
	// Parse command line flags
	flag.IntVar(&config.Port, "port", 9090, "Port to listen on")
	flag.StringVar(&config.Host, "host", "0.0.0.0", "Host to bind to")
	flag.StringVar(&config.Origin, "origin", "*", "CORS origin to allow")
	flag.StringVar(&config.Username, "username", "", "Basic auth username")
	flag.StringVar(&config.Password, "password", "", "Basic auth password")
	flag.BoolVar(&config.Verbose, "verbose", false, "Enable verbose logging")
	flag.Parse()

	startTime = time.Now()

	// Set up routes
	http.HandleFunc("/health", handleHealth)
	http.HandleFunc("/", handleProxy)

	// Start server
	addr := fmt.Sprintf("%s:%d", config.Host, config.Port)
	fmt.Printf(`
🚀 API Client CORS Proxy Server (Go) started!
   
   Listening on: http://%s
   Health check: http://%s/health
   
   Usage:
   - Add X-Target-URL header with the target API URL
   - Or use ?url=<target-url> query parameter
   
   %s
   %s

`, addr, addr,
		func() string {
			if config.Username != "" {
				return "🔒 Authentication enabled"
			}
			return "🔓 No authentication"
		}(),
		func() string {
			if config.Verbose {
				return "📝 Verbose logging enabled"
			}
			return ""
		}())

	log.Fatal(http.ListenAndServe(addr, nil))
}

func handleHealth(w http.ResponseWriter, r *http.Request) {
	setCORSHeaders(w, r)
	
	if r.Method == "OPTIONS" {
		w.WriteHeader(http.StatusNoContent)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(HealthResponse{
		Status:  "ok",
		Version: "1.0.0",
		Uptime:  time.Since(startTime).Seconds(),
	})
}

func handleProxy(w http.ResponseWriter, r *http.Request) {
	setCORSHeaders(w, r)
	
	if r.Method == "OPTIONS" {
		w.WriteHeader(http.StatusNoContent)
		return
	}

	// Check authentication
	if !checkAuth(r) {
		w.Header().Set("WWW-Authenticate", `Basic realm="Proxy"`)
		w.WriteHeader(http.StatusUnauthorized)
		w.Write([]byte("Unauthorized"))
		return
	}

	// Extract target URL
	targetURL := r.Header.Get("X-Target-URL")
	if targetURL == "" {
		targetURL = r.URL.Query().Get("url")
	}

	if targetURL == "" {
		w.WriteHeader(http.StatusBadRequest)
		w.Write([]byte("Missing target URL. Use X-Target-URL header or ?url= parameter"))
		return
	}

	// Create proxy request
	proxyReq, err := http.NewRequest(r.Method, targetURL, r.Body)
	if err != nil {
		sendError(w, "Invalid target URL", err)
		return
	}

	// Copy headers
	skipHeaders := map[string]bool{
		"Host":               true,
		"Connection":         true,
		"Proxy-Authorization": true,
		"X-Target-URL":       true,
		"Origin":             true,
		"Referer":            true,
	}

	for key, values := range r.Header {
		if !skipHeaders[key] {
			for _, value := range values {
				proxyReq.Header.Add(key, value)
			}
		}
	}

	// Log request if verbose
	if config.Verbose {
		log.Printf("[%s] %s %s", time.Now().Format("2006-01-02 15:04:05"), r.Method, targetURL)
	}

	// Make request
	client := &http.Client{
		Timeout: 30 * time.Second,
	}
	resp, err := client.Do(proxyReq)
	if err != nil {
		sendError(w, "Proxy error", err)
		if config.Verbose {
			log.Printf("[%s] %s %s -> Error: %v", time.Now().Format("2006-01-02 15:04:05"), r.Method, targetURL, err)
		}
		return
	}
	defer resp.Body.Close()

	// Log response if verbose
	if config.Verbose {
		log.Printf("[%s] %s %s -> %d", time.Now().Format("2006-01-02 15:04:05"), r.Method, targetURL, resp.StatusCode)
	}

	// Copy response headers
	for key, values := range resp.Header {
		if key != "Connection" && key != "Transfer-Encoding" {
			for _, value := range values {
				w.Header().Add(key, value)
			}
		}
	}

	// Set CORS headers (after copying response headers to override)
	setCORSHeaders(w, r)

	// Set status code
	w.WriteHeader(resp.StatusCode)

	// Copy response body
	io.Copy(w, resp.Body)
}

func setCORSHeaders(w http.ResponseWriter, r *http.Request) {
	origin := r.Header.Get("Origin")
	if origin == "" {
		origin = "*"
	}

	if config.Origin == "*" {
		w.Header().Set("Access-Control-Allow-Origin", origin)
	} else {
		w.Header().Set("Access-Control-Allow-Origin", config.Origin)
	}

	w.Header().Set("Access-Control-Allow-Methods", "GET, POST, PUT, DELETE, PATCH, OPTIONS, HEAD")
	w.Header().Set("Access-Control-Allow-Headers", "*")
	w.Header().Set("Access-Control-Allow-Credentials", "true")
	w.Header().Set("Access-Control-Max-Age", "86400")
}

func checkAuth(r *http.Request) bool {
	if config.Username == "" || config.Password == "" {
		return true
	}

	auth := r.Header.Get("Proxy-Authorization")
	if auth == "" {
		auth = r.Header.Get("Authorization")
	}

	if auth == "" {
		return false
	}

	parts := strings.SplitN(auth, " ", 2)
	if len(parts) != 2 || parts[0] != "Basic" {
		return false
	}

	decoded, err := base64.StdEncoding.DecodeString(parts[1])
	if err != nil {
		return false
	}

	credentials := strings.SplitN(string(decoded), ":", 2)
	if len(credentials) != 2 {
		return false
	}

	return credentials[0] == config.Username && credentials[1] == config.Password
}

func sendError(w http.ResponseWriter, message string, err error) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusInternalServerError)
	json.NewEncoder(w).Encode(ErrorResponse{
		Error:   message,
		Message: err.Error(),
	})
}